File: src/main/java/com/example/coupon/coupon/infrastructure/repository/CouponRepository.java
package com.example.coupon.coupon.infrastructure.repository;

import com.example.coupon.coupon.domain.entity.Coupon;

import java.util.Optional;

public interface CouponRepository {

    Coupon save(Coupon coupon);

    Optional<Coupon> getTodayEventCoupon();
}

File: src/main/kotlin/com/example/couponservice/domain/CouponRepository.kt
package com.example.couponservice.domain

import jakarta.persistence.LockModeType
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Query
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.repository.query.Param

interface CouponRepository: JpaRepository<Coupon, Long> {
    fun findByName(name: String): Coupon?

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("select c from Coupon c where c.name = :name")
    fun findByNameWithLock(@Param("name") name: String): Coupon?
}
